singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -647,6 +651,491 @@ class PTESolverRhoT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
+        const Real rho_min = eos[m].RhoPmin(Tguess);
+        rho[m] = rhobar[m] / vfrac[m];
+        if (rho[m] < rho_min) {
+          rho_fail = true;
+          Tguess *= Tfactor;
+          break;
+        }
+      }
+      if (!rho_fail) break;
+    }
+
+    if (rho_fail) {
+      printf("rho < rho_min in PTE initialization!  Solver may not converge.");
+    }
+    // set the temperature normalization
+    Tnorm = Tguess;
+    // calculate u normalization as internal energy guess
+    u_total = 0.0;
+    for (int m = 0; m < nmat; m++) {
+      // scaled initial guess for temperature is just 1
+      temp[m] = 1.0;
+      sie[m] = eos[m].InternalEnergyFromDensityTemperature(rho[m], Tguess);
+      u_total += sie[m] * rho[m];
+    }
+
+    // note the scaling of the material internal energy densities
+    for (int m = 0; m < nmat; ++m)
+      u[m] = sie[m] * rhobar[m] / u_total;
+    Residual();
+    // Set the current guess for the equilibrium temperature.  Note that this is already
+    // scaled.
+    Tequil = temp[0];
+    return ResidualNorm();
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  void Residual() const {
+    Real vsum = 0.0;
+    for (int m = 0; m < nmat; ++m) {
+      vsum += vfrac[m];
+      residual[m] = Pequil - press[m];
+    }
+    residual[neq-1] = vfrac_total - vsum;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  bool CheckPTE() const {
+    using namespace mix_params;
+    Real mean_p = vfrac[0] * press[0];
+    Real error_p = 0.0;
+    for (int m = 1; m < nmat; ++m) {
+      mean_p += vfrac[m] * press[m];
+      error_p += residual[m + 1] * residual[m + 1];
+    }
+    error_p = std::sqrt(error_p);
+    Real error_v = std::abs(residual[neq-1]);
+    // Check for convergence
+    bool converged_p = (error_p < pte_rel_tolerance_p * std::abs(mean_p) ||
+                        error_p < pte_abs_tolerance_p);
+    bool converged_v = (error_v < pte_rel_tolerance_e || error_v < pte_abs_tolerance_e);
+    return converged_p && converged_v;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  void Jacobian() const {
+    using namespace mix_params;
+    for (int m = 0; m < nmat; m++) {
+      //////////////////////////////
+      // perturb volume fractions
+      //////////////////////////////
+      Real dv = (vfrac[m] < 0.5 ? 1.0 : -1.0) * vfrac[m] * derivative_eps;
+      const Real vf_pert = vfrac[m] + dv;
+      const Real rho_pert = rhobar[m] / vf_pert;
+
+      Real p_pert{};
+      Real e_pert = eos[m].InternalEnergyFromDensityTemperature(rho_pert, Tnorm * Tequil, Cache[m]);
+      if (eos[m].PreferredInput() ==
+          (thermalqs::density | thermalqs::specific_internal_energy)) {
+        p_pert = eos[m].PressureFromDensityInternalEnergy(rho_pert, e_pert, Cache[m]) /
+                 u_total;
+      } else if (eos[m].PreferredInput() ==
+                 (thermalqs::density | thermalqs::temperature)) {
+        p_pert =
+            eos[m].PressureFromDensityTemperature(rho_pert, Tnorm * Tequil, Cache[m]) /
+            u_total;
+      }
```

Comment:
In the fixed T solver, I enforced the using T since it is known to be correct. In this case, I left the logic to use the preferred eos method for the lookup.